federation/pkg/federation-controller/ingress/ingress_controller.go

```diff
@@ -218,40 +338,240 @@ func (ic *IngressController) deliverIngress(ingress types.NamespacedName, delay
+	glog.V(4).Infof("Reconciling ConfigMap %q in cluster %q", configMapNsName, cluster.Name)
+
+	clusterIngressUID, clusterIngressUIDExists := cluster.ObjectMeta.Annotations[uidAnnotationKey]
+	configMapUID, ok := configMap.Data[uidKey]
+
+	if !ok {
+		glog.Errorf("Warning: ConfigMap %q in cluster %q does not contain data key %q.  Therefore it cannot become the master.", configMapNsName, cluster.Name, uidKey)
+	}
+
+	if !clusterIngressUIDExists || clusterIngressUID == "" {
+		ic.updateClusterIngressUIDToMasters(cluster, configMapUID) // Second argument is the fallback, in case this is the only cluster, in which case it becomes the master
+		return
+	}
+	if configMapUID != clusterIngressUID { // An update is required
+		glog.V(4).Infof("Ingress UID update is required: configMapUID %q not equal to clusterIngressUID %q", configMapUID, clusterIngressUID)
+		configMap.Data[uidKey] = clusterIngressUID
+		operations := []util.FederatedOperation{{
+			Type:        util.OperationTypeUpdate,
+			Obj:         configMap,
+			ClusterName: cluster.Name,
+		}}
+		glog.V(4).Infof("Calling federatedConfigMapUpdater.Update() - operations: %v", operations)
+		err := ic.federatedConfigMapUpdater.Update(operations, ic.updateTimeout)
+		if err != nil {
+			configMapName := types.NamespacedName{Name: configMap.Name, Namespace: configMap.Namespace}
+			glog.Errorf("Failed to execute update of ConfigMap %q on cluster %q: %v", configMapName, cluster.Name, err)
+			ic.configMapDeliverer.DeliverAfter(cluster.Name, nil, ic.configMapReviewDelay)
+		}
+	} else {
+		glog.V(4).Infof("Ingress UID update is not required: configMapUID %q is equal to clusterIngressUID %q", configMapUID, clusterIngressUID)
+	}
+}
+
+/*
+  getMasterCluster returns the cluster which is the elected master w.r.t. ingress UID, and it's ingress UID.
+  If there is no elected master cluster, an error is returned.
+  All other clusters must use the ingress UID of the elected master.
+*/
+func (ic *IngressController) getMasterCluster() (master *federation_api.Cluster, ingressUID string, err error) {
+	clusters, err := ic.configMapFederatedInformer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get cluster list: %v", err)
+		return nil, "", err
+	}
+
+	for _, c := range clusters {
+		UID, exists := c.ObjectMeta.Annotations[uidAnnotationKey]
+		if exists && UID != "" { // Found the master cluster
+			glog.V(4).Infof("Found master cluster %q with annotation %q=%q", c.Name, uidAnnotationKey, UID)
+			return c, UID, nil
+		}
+	}
+	return nil, "", fmt.Errorf("Failed to find master cluster with annotation %q", uidAnnotationKey)
+}
+
+/*
+  updateClusterIngressUIDToMasters takes the ingress UID annotation on the master cluster and applies it to cluster.
+  If there is no master cluster, then fallbackUID is used (and hence this cluster becomes the master).
+*/
+func (ic *IngressController) updateClusterIngressUIDToMasters(cluster *federation_api.Cluster, fallbackUID string) {
```

Comment:
This ideally needs to be done in parallel via a worker queue.  Will do in a followup PR.